PYTHON ASSIGNMENT
                                                      

 Python is a versatile, high-level, general-purpose programming language known for its readability and ease of use, widely used in web development, data science, and machine learning. 

Section 1: Python Operators and Control Structures (5 questions)

1. Question: Write a Python program to calculate the sum of all numbers in a list using a for loop. 

   Sample Input: numbers = [1, 2, 3, 4, 5]
   
   Sample Output: 15


In [5]:
numbers = [1, 2, 3, 4, 5]
sum=0
for i in numbers:
    sum=s+i
print(sum)

15


2. Question: What is the output of the following code?

    Code:
   
      x = 5  

      y = 2
    print(x ** y)
  
    Sample Output: 25


In [10]:
x=5
y=2
print(x**y) 

25


 Explanation 
     
     ** -> It is a exponential operator. So, it retruns 5**2=25    


 

3. Question: Write a Python program to check if a number is a palindrome using a while loop.


      Sample Input: num = 12321


      Sample Output: True
   

In [4]:
num=12321
n=num
res=0
rem=0
while num>0:
    rem=num%10
    res=(res*10)+rem
    num=num//10
if n==res:
    print("true")
else :
    print("false")

true


4  Question: What is the output of the following code?

  Code:
 
    fruits = ['apple', 'banana', 'cherry']
   
    print(fruits[-1])
    
    Sample Output: cherry
 

In [5]:
fruits=['apple','banana','cherry']
print(fruits[-1])

cherry


Explanation
     
    fruits[-1] is index position of cherry

5. Question: Write a Python program to calculate the average of all numbers in a list using a for loop.

    Sample Input: numbers = [1, 2, 3, 4, 5]

     Sample Output: 3.0




In [9]:
numbers=[1,2,3,4,5]
s=0
for i in numbers:
    s=i+s
avg=s/len(numbers)
print(avg)

3.0


Section 2: Lists and Dictionaries (5 questions)

6. Question: Write a Python program to create a dictionary with student names as keys and their grades as values.

    Sample Input: students = ['John', 'Alice', 'Bob'], grades = [90, 85, 95]
   
    Sample Output: {'John': 90, 'Alice': 85, 'Bob': 95}

In [12]:
 students = ['John', 'Alice', 'Bob']
 grades = [90, 85, 95]
d={}
for i in range(len(students)):
    d[students[i]]=grades[i]
print(d)

{'John': 90, 'Alice': 85, 'Bob': 95}


7. Question: What is the output of the following code?
 
   Code:

    numbers = [1, 2, 3, 4, 5]
   
    print(numbers[1:3])
   
    Sample Output: [2, 3]


In [1]:
numbers = [1, 2, 3, 4, 5]
print(numbers[1:3])

[2, 3]


Explanation

    numbers[1:3] is list slicing.
    
    It extracts elements starting from index 1 up to (but not including) index 3.

8. Question: Write a Python program to sort a list of numbers in ascending order using the sorted() function.

     Sample Input: numbers = [4, 2, 9, 6, 5]

     Sample Output: [2, 4, 5, 6, 9]


In [2]:
 numbers = [4, 2, 9, 6, 5]
 print(sorted(numbers))

[2, 4, 5, 6, 9]


Explanation
    
     sorted() function is used to arrange given list in ascending order

9. Question: What is the output of the following code?

   Code:

      fruits = {'apple': 5, 'banana': 10, 'cherry': 15}
   
      print(fruits.get('banana'))

      Sample Output: 10




In [3]:
fruits = {'apple': 5, 'banana': 10, 'cherry': 15}

print(fruits.get('banana'))

10


Explanation

    In fruits dictionary, we are retrieving a value of key(banana) by using get() method.

10. Question: Write a Python program to create a list of squares of all numbers from 1 to 10 using a list comprehension.

    Sample Output: [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]




In [5]:
n=[i**2 for i in range(1,11)]
print(n)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Explanation

     List comprehension is a concise way to create lists in Python using a single line of code. It replaces loops for generating lists, making the code more readable and efficient.

     new_list = [expression(element) 'for' element 'in' iterable_object 'if' condition ]
     


Section 3: Advanced Topics (5 questions)

11. Question: Write a Python program to calculate the factorial of a number using recursion.

    Sample Input: num = 5

    Sample Output: 120


In [2]:
def factorial(n):
    if n == 0 or n == 1:
        return 1
    return n * factorial(n - 1)

num = 5
print(factorial(num))

120


12. Question: What is the output of the following code?

    Code:
   
     numbers = [1, 2, 3, 4, 5]
    
     print(list(reversed(numbers)))
     
     Sample Output: [5, 4, 3, 2, 1]

In [3]:
numbers = [1, 2, 3, 4, 5]

print(list(reversed(numbers)))

[5, 4, 3, 2, 1]


Explanation:
      
       reversed(numbers) returns an iterator that produces elements in reverse order.
       list(reversed(numbers)) converts the iterator into a list.

13. Question: Write a Python program to create a generator that yields the Fibonacci sequence up to the nth term.

    Sample Input: n = 10
    
    Sample Output: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [26]:
def fibonacci(n):
    a, b = 0, 1
    for i in range(n):
        yield a
        a, b = b, a + b

n = 10
print(list(fibonacci(n)))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


Explanation:
   
    yield is used to create a generator instead of storing all values in memory.
    a, b = 0, 1 initializes the first two numbers.
    The loop runs n times, yielding a each time and updating a and b.

14. Question: What is the output of the following code?

     Code:

       import math

       print(math.pi)

       Sample Output: 3.14159...




In [2]:
import math

print(math.pi)

3.141592653589793


Explanation

      The math module is imported and math.pi is a constant in the math module that represents the mathematical value of π (pi).
      The value of math.pi is approximately 3.141592653589793.

15. Question: Write a Python program to print the inverted star pyramid of height 5.


In [19]:
for i in range(5,0,-1):
    print(" "*(5-i)+"* "*i)

* * * * * 
 * * * * 
  * * * 
   * * 
    * 


Explanation:
      
      The for loop runs from height(5) down to 1, decreasing in each iteration.
      " " * (5-i) adds leading spaces to align the pyramid.
      "* " * i prints decreasing stars in each row.